drivers/flash/flash_esp32.c

```diff
@@ -0,0 +1,622 @@
+
+	int rc = flash_esp32_wait_cmd_done(cfg->controller);
+
+	if (rc != 0) {
+		return rc;
+	}
+
+	/* make sure the flash is ready for writing */
+	while (ESP_ROM_SPIFLASH_WRENABLE_FLAG != (flash_status & ESP_ROM_SPIFLASH_WRENABLE_FLAG)) {
+		read_status(dev, &flash_status);
+	}
+
+	return 0;
+}
+
+static int program_page(const struct device *dev, uint32_t spi_addr,
+			uint32_t *addr_source, int32_t byte_length)
+{
+	const struct flash_esp32_dev_config *const cfg = DEV_CFG(dev);
+	spi_dev_t *hw = DEV_CFG(dev)->controller;
+
+	/* check 4byte alignment */
+	if ((byte_length & 0x3) != 0) {
+		return -EINVAL;
+	}
+
+	/* check if write in one page */
+	if ((cfg->chip->page_size) < ((spi_addr % (cfg->chip->page_size)) + byte_length)) {
+		return -EINVAL;
+	}
+
+	wait_idle(dev);
+
+	uint32_t addr;
+	uint32_t prog_len;
+
+	while (byte_length > 0) {
+		if (write_protect(dev, false) != 0) {
+			return -EINVAL;
+		}
+
+		addr = spi_addr & ADDRESS_MASK_24BIT;
+
+		if (byte_length >= ESP_ROM_SPIFLASH_BUFF_BYTE_WRITE_NUM) {
+			addr |= ESP_ROM_SPIFLASH_BUFF_BYTE_WRITE_NUM << ESP_ROM_SPIFLASH_BYTES_LEN;
+			prog_len = (uint32_t)ESP_ROM_SPIFLASH_BUFF_BYTE_WRITE_NUM;
+			spi_flash_ll_set_address(hw, addr);
+			spi_flash_ll_program_page(hw, addr_source, prog_len);
+			byte_length -= ESP_ROM_SPIFLASH_BUFF_BYTE_WRITE_NUM;
+			spi_addr += ESP_ROM_SPIFLASH_BUFF_BYTE_WRITE_NUM;
+		} else {
+			addr |= byte_length << ESP_ROM_SPIFLASH_BYTES_LEN;
+			prog_len = (uint32_t)byte_length;
+			spi_flash_ll_set_address(hw, addr);
+			spi_flash_ll_program_page(hw, addr_source, prog_len);
+			byte_length = 0;
+		}
+
+		int rc = flash_esp32_wait_cmd_done(hw);
+
+		if (rc != 0) {
+			return rc;
+		}
+
+		wait_idle(dev);
+	}
+
+	return 0;
+}
+
+static int flash_esp32_write(const struct device *dev,
+			     off_t address,
+			     const void *buffer,
+			     size_t length)
+{
+	const struct flash_esp32_dev_config *const cfg = DEV_CFG(dev);
+
+	uint32_t page_size;
+	uint32_t prog_len, prog_num;
+	int rc = 0;
+
+	flash_esp32_sem_take(dev);
+
+	set_write_options(dev);
+
+	/* check program size */
+	if ((address + length) > (cfg->chip->chip_size)) {
+		rc = -EINVAL;
+		goto out;
+	}
+
+	page_size = cfg->chip->page_size;
+	prog_len = page_size - (address % page_size);
+	if (length < prog_len) {
+		rc = program_page(dev, address, (uint32_t *)buffer, length);
+		if (rc) {
+			goto out;
+		}
+	} else {
+		rc = program_page(dev, address, (uint32_t *)buffer, prog_len);
+
+		if (rc) {
+			goto out;
+		}
+
+		/* whole page program */
+		prog_num = (length - prog_len) / page_size;
+		for (uint8_t i = 0; i < prog_num; ++i) {
+			rc = program_page(dev,
+					  address + prog_len,
+					  (uint32_t *)buffer + (prog_len >> 2),
+					  page_size);
+
+			if (rc) {
+				goto out;
+			}
+
+			prog_len += page_size;
+		}
+
+		/* remain parts to program */
+		rc = program_page(dev,
+				  address + prog_len,
+				  (uint32_t *)buffer + (prog_len >> 2),
+				  length - prog_len);
+		if (rc) {
```

Comment:
Is this if necessary?

Reply:
I've added a `LOG_ERR` message to it. I guess it is more meaningful now.